### PR TITLE
Colab to git

### PR DESCRIPTION
**Pull request by agpar**
This PR moves a lot of the code that should be common to the approaches into the repo.

* Added files in `machine_learning` that hold relevant functionality from colab. Removed files that had out of date versions of this code.

* Added files in `machine_learning/text_embedders` that take responsibility for doing either BOW or GLOVE encoding.
* `get_model_data` now takes as argument a `TextEmbedder` and is ignorant about whether GLOVE or BOW embedding is done.

Overall, up until the point where the GAT model is set up, the two approaches should be identical.

### CODE REVIEW

File: machine_learning/text_embedders/glove_embedder.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        # https://stackoverflow.com/questions/49239941/what-is-unk-in-the-pretrained-glove-vector-files-e-g-glove-6b-50d-txt
+        avg_vec = np.mean(vectors[1:].reshape((-1, 50)), axis=0)
+        unknown_word = "[unk]" # Note: "unk" and "<unk>" are actually words in GloVe
+
+        word2idx[unknown_word] = len(words)
+        words.append(unknown_word)
+        vectors.append(avg_vec)
+
+        # Save the results
+        vectors = bcolz.carray(vectors[1:].reshape((-1, 50)), rootdir=f'{self.glove_path}/6B.{self.dim}.dat', mode='w')
```
Comment (jpngfile) [inline machine_learning/text_embedders/glove_embedder.py:53]: These lines should use `self.dim` instead of 50

**Review by jpngfile — APPROVED** [review]
I'm fine with these changes. It should be acknowledged that the model does have some changes to accommodate different input and output types. Namely:
- Subreddit classification has 3 classes instead of 2, so the model output needs 3 nodes and precision and recall can't be used
- Glove requires an embedding layer, and needs to pass in `offsets` as well as the word vectors. To create the embedding layer, the model needs the list of word vectors
- The data returned has different names. Bag-of-words uses `bag_of_words`, and GloVe uses both `embedding` and `offsets`

We may want to make it more evident that the types of inputs and outputs are not completely interchangeable.

**Comment by agpar** [comment]
> Subreddit classification has 3 classes instead of 2, so the model output needs 3 nodes and precision and recall can't be used

I haven't quite "unified" the model at the end, recognizing that this is where things start to diverge the most.


> Glove requires an embedding layer, and needs to pass in offsets as well as the word vectors. To create the embedding layer, the model needs the list of word vectors

Yes, this is one of the differences I noticed. The word vectors are saved on GloveEmbedder to be accessed later on by the model

> The data returned has different names. Bag-of-words uses bag_of_words, and GloVe uses both embedding and offsets.

You can see in `machine_learning/batch_combiner.py` that I've standardized the outputs.

**Comment by agpar** [comment]
I have some other changes I made to the models. For instance, it's not necessary to load all vectors into the GPU before training, you can load one batch at a time (allowing larger data sets).

I haven't decided on a good way to share these changes though.

**Pull request merged**